### ISSUE TITLE
Inquiry about Sparse IPA Inputs in AlphaFold Implementation

### ISSUE THREAD

**Issue by guyujun** [mjendrusch/salad#3]
I hope this message finds you well. I have a question regarding the input handling in the sparse IPA component of the AlphaFold implementation (appears to be a pre-release version).

Could you please clarify:

Whether the noised inputs to sparse IPA use the atom14 representation or specifically backbone atoms (N, CA, C)?
How the side-chain atoms of residues are constructed in this context?
I would greatly appreciate any insights you could share to help me better understand this implementation detail.

Thank you for your time and expertise.

**Comment by mjendrusch** [comment]
Hi! I'm not sure what part of the code you're referring to, but if it's the SparseIPA implementation in `salad/modules/geometric.py` (`SparseInvariantPointAttention`) in `salad/modules/noise_schedule_benchmark.py`, this takes as input frames constructed from the N, CA and C atoms, same as IPA in AlphaFold.

If you are asking about the `pos_noised` input to `StructureDiffusion` and `StructureDiffusionInference` in `salad/modules/noise_schedule_benchmark.py`, this has the N, CA, C, O atoms for each residue, plus idealized CB positions and an additional `config.augment_size` number of learned pseudo-atom positions.
The side-chains are not represented directly in the denoising module (`DiffusionStack`). Instead, the `AADecoder` module predicts the amino acid identity of each residue, which is then used to predict side-chain dihedral angles. These angles are then used to construct the side-chain positions for each residue.

This is the part in `noise_schedule_benchmark.py` where the `AADecoder` is used to predict an amino acid sequence which is then used to construct side-chain coordinates. The output all-atom positions including side-chains are returned in atom14 format in `result["atom_pos"]`:
https://github.com/mjendrusch/salad/blob/2114bdabc710e2ab3f0d2a8b56c80793c7da77b1/salad/modules/noise_schedule_benchmark.py#L913-L929

Hope that helps!

Let me know if there's any particular part of the code that you would like to know more about – if you point me towards some specific lines in the code base that would help me narrow down what I should elaborate on.

**Comment by guyujun** [comment]
> Hi! I'm not sure what part of the code you're referring to, but if it's the SparseIPA implementation in `salad/modules/geometric.py` (`SparseInvariantPointAttention`) in `salad/modules/noise_schedule_benchmark.py`, this takes as input frames constructed from the N, CA and C atoms, same as IPA in AlphaFold.
> 
> If you are asking about the `pos_noised` input to `StructureDiffusion` and `StructureDiffusionInference` in `salad/modules/noise_schedule_benchmark.py`, this has the N, CA, C, O atoms for each residue, plus idealized CB positions and an additional `config.augment_size` number of learned pseudo-atom positions. The side-chains are not represented directly in the denoising module (`DiffusionStack`). Instead, the `AADecoder` module predicts the amino acid identity of each residue, which is then used to predict side-chain dihedral angles. These angles are then used to construct the side-chain positions for each residue.
> 
> This is the part in `noise_schedule_benchmark.py` where the `AADecoder` is used to predict an amino acid sequence which is then used to construct side-chain coordinates. The output all-atom positions including side-chains are returned in atom14 format in `result["atom_pos"]`:
> 
> [salad/salad/modules/noise_schedule_benchmark.py](https://github.com/mjendrusch/salad/blob/2114bdabc710e2ab3f0d2a8b56c80793c7da77b1/salad/modules/noise_schedule_benchmark.py#L913-L929)
> 
> Lines 913 to 929 in [2114bda](/mjendrusch/salad/commit/2114bdabc710e2ab3f0d2a8b56c80793c7da77b1)
> 
>  # generate all-atom positions using predicted 
>  # side-chain torsion angles: 
>  aatype = data["aa_gt"] 
>  if predict_aa: 
>      if c.sample_aa: 
>          aatype = aa_decoder.sample( 
>              data["aa_gt"], local, pos, resi, chain, batch, mask) 
>      else: 
>          aatype = jnp.argmax(aa_logits, axis=-1) 
>  if "aa_condition" in data: 
>      aatype = jnp.where(data["aa_condition"] != 20, data["aa_condition"], aatype) 
>  result["aatype"] = aatype 
>  raw_angles, angles, angle_pos = get_angle_positions( 
>      aatype, local, pos) 
>  result["raw_angles"] = raw_angles 
>  result["angles"] = angles 
>  result["atom_pos"] = angle_pos 
> Hope that helps!
> 
> Let me know if there's any particular part of the code that you would like to know more about – if you point me towards some specific lines in the code base that would help me narrow down what I should elaborate on.

Thank you for your response. I fully understand how SALAD operates. Additionally, I have an extra question. I tried using SALAD's attention sparsification approach, but the training speed in JAX doesn't match that of standard attention. Have you encountered a similar phenomenon? Why doesn't this sparsification method attempt to be used in Triangle?

**Comment by guyujun** [comment]
`class Attention(hk.Module):
  def __init__(self, config, global_config, output_dim, name='attention'):
    super().__init__(name=name)

    self.config = config
    self.global_config = global_config
    self.output_dim = output_dim

  def __call__(self, q_data, m_data, padding_mask, nonbatched_bias=None, neighbours=None):
    """Builds Attention module.

    Arguments:
      q_data: A tensor of queries, shape [batch_size, N_queries, q_channels].
      m_data: A tensor of memories from which the keys and values are
        projected, shape [batch_size, N_keys, m_channels].
      padding_mask: A bias for the attention, shape [batch_size, 1, 1, N_keys]. for bhqk+b11k
      nonbatched_bias: Shared bias, shape [N_head, N_queries, N_keys].
      neighbours: A tensor of neighbours, shape [N_queries, N_neighbours].
    Returns:
      A float32 tensor of shape [batch_size, N_queries, output_dim].
    """
    # Sensible default for when the config keys are missing
    key_dim = self.config.get('key_dim', int(q_data.shape[-1]))
    value_dim = self.config.get('value_dim', int(m_data.shape[-1]))
    num_head = self.config.num_head
    assert key_dim % num_head == 0
    assert value_dim % num_head == 0
    key_dim = key_dim // num_head
    value_dim = value_dim // num_head
    dtype = q_data.dtype

    if self.config.use_q_bias:
      # q has bias
      q_weights = hk.get_parameter(
          'query_w', shape=(q_data.shape[-1], num_head, key_dim),
          dtype=q_data.dtype,
          init=glorot_uniform())
      b_weights = hk.get_parameter(
          'query_b', shape=(num_head, key_dim),
          dtype=q_data.dtype,
          init=hk.initializers.Constant(0.0))
      q = (jnp.einsum('bqa,ahc->bqhc', q_data, q_weights) + b_weights) * key_dim**(-0.5)
      
    else:
      # q has no-bias
      q_weights = hk.get_parameter(
          'query_w', shape=(q_data.shape[-1], num_head, key_dim),
          dtype=q_data.dtype,
          init=glorot_uniform())
      q = jnp.einsum('bqa,ahc->bqhc', q_data, q_weights) * key_dim**(-0.5)
    
    # [batch_size, N_queries, head, q_channels], LinearNoBias:
    k_weights = hk.get_parameter(
        'key_w', shape=(m_data.shape[-1], num_head, key_dim),
        dtype=q_data.dtype,
        init=glorot_uniform())
    v_weights = hk.get_parameter(
        'value_w', shape=(m_data.shape[-1], num_head, value_dim),
        dtype=q_data.dtype,
        init=glorot_uniform())
    
    # "ka,ahc->khc" 等价于矩阵乘法: ka @ ahc -> khc.
    k = jnp.einsum('bka,ahc->bkhc', m_data, k_weights)
    v = jnp.einsum('bka,ahc->bkhc', m_data, v_weights)
    
    # get attention-score:
    if neighbours is None:
      # RESHAPE bhqk方便加bias. 'bqhc,bkhc->bqkh'
      logits = jnp.einsum('bqhc,bkhc->bhqk', q, k) + padding_mask # bhqk + (b,1,1,k)
    else:
      # knn切片+广播: k_broadcast_slice.shape = (b, k, n, h, c)
      k_broadcast_slice = k[:,neighbours,:,:].astype(dtype)
      # neighbours.shape = (q,n)
      padding_mask = padding_mask[...,:neighbours.shape[-1]].astype(dtype)  # slice to (b,1,1,n)
      # "bhic,bhijc->bhij"
      logits = jnp.einsum('bihc,bijhc->bhij', q, k_broadcast_slice) + padding_mask # bhkn + (b,1,1,n)

    if nonbatched_bias is not None:
      if neighbours is None:
        logits += jnp.expand_dims(nonbatched_bias, axis=0)               # bhqk + (1,hqk)
      else:
        row_indices = jnp.arange(q.shape[1])[:, None]                    # (1,q)
        nonbatched_bias = nonbatched_bias[:, row_indices, neighbours].astype(dtype)    # hkn
        logits += jnp.expand_dims(nonbatched_bias, axis=0).astype(dtype) # bhkn + (1,hkn)

    weights = stable_softmax(logits)
    if neighbours is None:
      weighted_avg = jnp.einsum('bhqk,bkhc->bqhc', weights, v)
    else:
      v_broadcast_slice = v[:,neighbours,:,:].astype(dtype) # (b, k, n, h, c)
      weighted_avg = jnp.einsum('bhij,bijhc->bihc', weights, v_broadcast_slice)

    if self.global_config.zero_init:
      init = hk.initializers.Constant(0.0)
    else:
      init = glorot_uniform()

    if self.config.gating:
      gating_weights = hk.get_parameter(
          'gating_w',
          shape=(q_data.shape[-1], num_head, value_dim),
          dtype=q_data.dtype,
          init=hk.initializers.Constant(0.0))
      gate_values = jnp.einsum('bqc, chv->bqhv', q_data, gating_weights)
      gate_values = jax.nn.sigmoid(gate_values)
      weighted_avg *= gate_values

    o_weights = hk.get_parameter(
        'output_w', shape=(num_head, value_dim, self.output_dim),
        dtype=q_data.dtype,
        init=init)

    output = jnp.einsum('bqhc,hco->bqo', weighted_avg, o_weights) # + o_bias

    return output`

here is my implement.

**Comment by mjendrusch** [comment]
Without benchmarking / profiling the code I cannot really tell what is going on.
However, if I may take a guess: both computing the neighbour list and constructing the `*_broadcast_slice` arrays is an additional cost compared to regular attention, which does not explicitly construct key and value arrays of shape `(N_queries, N_queries, num_head, key_dim)`. I would expect this cost to out-weight the benefit of sparsity at small `N_queries` or large `N_neighbours`. 
What `N_queries` and `N_neighbours` are you using?
For `salad` I was using `N_queries = 1024` and `N_neighbours` between 32 and 64 for different models. At those sizes `SparseInvariantPointAttention` outperformed regular `InvariantPointAttention` in terms of runtime.
To figure out what exactly is causing the slow-down you might want to use the jax profiler, though.

Regarding using a neighbour list for `Triangle*` operations, these operate on pair features. You could restrict one or both of the pair indices to a neighbour list for each residue to get either `N_queries × N_queries × N_neighbours` operations (update all pair features, attending to only `N_neighbour` other pairs for each amino acid pair), or `N_queries × N_neighbours × N_neighbours` operations (updating only the nearest neighbour pair features for each amino acid, attending over only those neighbours). I believe Wang _et al._ did something along those lines for Proteus (https://www.biorxiv.org/content/10.1101/2024.02.10.579791v3).

I cannot really comment on how well that would work, as I haven't tried using sparse `TriangleMultiplication` / `TriangleAttention` inside `salad`. The reason for that I wanted to avoid using pair features and O(N³) operations on top of them to save on runtime.

**Comment by guyujun** [comment]
> Without benchmarking / profiling the code I cannot really tell what is going on. However, if I may take a guess: both computing the neighbour list and constructing the `*_broadcast_slice` arrays is an additional cost compared to regular attention, which does not explicitly construct key and value arrays of shape `(N_queries, N_queries, num_head, key_dim)`. I would expect this cost to out-weight the benefit of sparsity at small `N_queries` or large `N_neighbours`. What `N_queries` and `N_neighbours` are you using? For `salad` I was using `N_queries = 1024` and `N_neighbours` between 32 and 64 for different models. At those sizes `SparseInvariantPointAttention` outperformed regular `InvariantPointAttention` in terms of runtime. To figure out what exactly is causing the slow-down you might want to use the jax profiler, though.
> 
> Regarding using a neighbour list for `Triangle*` operations, these operate on pair features. You could restrict one or both of the pair indices to a neighbour list for each residue to get either `N_queries × N_queries × N_neighbours` operations (update all pair features, attending to only `N_neighbour` other pairs for each amino acid pair), or `N_queries × N_neighbours × N_neighbours` operations (updating only the nearest neighbour pair features for each amino acid, attending over only those neighbours). I believe Wang _et al._ did something along those lines for Proteus (https://www.biorxiv.org/content/10.1101/2024.02.10.579791v3).
> 
> I cannot really comment on how well that would work, as I haven't tried using sparse `TriangleMultiplication` / `TriangleAttention` inside `salad`. The reason for that I wanted to avoid using pair features and O(N³) operations on top of them to save on runtime.

would you please to show more details about the dimension/shape of the Q K V and attention logits arrays?

In my code, the shape of `*_broadcast_slice` is `(1, N_queries, N_neighbors, num_head, channel_dim)`. I’m concerned that something might go wrong. Is there a potential issue I should be aware of?

I was using `N_queries` = 128 and `N_neighbours` between 16 and 32,  in my test it works worse than the standard Attention.

```
std:                                                                         PMAP Time taken: 0.3659043312072754 seconds
knn_attention for row Attention:                         PMAP Time taken: 0.41559743881225586 seconds
knn_attention for triangleAttention:                   PMAP Time taken: 1.517237901687622 seconds
```
